src/inc/Lfo.h

```diff
@@ -0,0 +1,88 @@
+#if !defined(__Lfo_hdr__)
+#define __Lfo_hdr__
+
+#define _USE_MATH_DEFINES
+#include <math.h>
+
+#include "ErrorDef.h"
+#include "Synthesis.h"
+#include "RingBuffer.h"
+
+class CLfo
+{
+public:
+    CLfo(float fModFreq, float fSampleRate)
+    {
+        m_pCRingBuffer = 0;
+        m_fRead = 0;
+        m_fModFreq = fModFreq;
+        m_fSampleRate = fSampleRate;
+        if (m_fModFreq == 0) {
+            m_iBufferLength = 1;
+        } else {
+            m_iBufferLength = static_cast<int>(1 / m_fModFreq);
+        }
+        m_pCRingBuffer = new CRingBuffer<float>(m_iBufferLength);
+    }
+    
+    ~CLfo()
+    {
+        delete m_pCRingBuffer;
+        m_pCRingBuffer = 0;
+    }
+    
+    Error_t setParam(float fParamValue)
+    {
+        m_fModFreq = fParamValue;
+        float *pSine = new float [m_iBufferLength];
+        
+        if (m_fModFreq == 0) {
+            m_iBufferLength = 1;
+        }
+        else {
+            m_iBufferLength = static_cast<int>(1 / m_fModFreq);
+        }
+        
+        CSynthesis::generateSine(pSine, m_fModFreq * m_fSampleRate, m_fSampleRate, m_iBufferLength);
+        m_pCRingBuffer-> putPostInc(pSine, m_iBufferLength);
+        delete [] pSine;
+        
+        return Error_t::kNoError;
+    }
+    
+    //return current values
+    float returnLfoVal()
+    {
+        float fLfo = m_pCRingBuffer->get(m_fRead);
+        
+        m_fRead = m_fRead + 1;
+        
+        if (m_fRead >= m_iBufferLength) {
+            m_fRead -= m_iBufferLength;
+        }
+        
+        return fLfo;
```

Comment:
Not sure if this was intentional or not, but there doesn't appear to be any fractional indexing occuring with `m_pCRingBuffer->get(m_fRead)` since `m_fRead` will always be an integer

Reply:
```
Error_t CVibrato::process(float **ppfInputBuffer, float **ppfOutputBuffer, int iNumberOfFrames) {
    float fModAmp, fDelay = 0;
    for(int i = 0; i < m_iNumChannels; i++) {
        for(int j = 0; j < iNumberOfFrames; j++) {
            fModAmp = m_pCLfo -> returnLfoVal();
            fDelay = 1 + m_fDelayInS * m_fSampleRateInHz + m_fModWidthInS * m_fSampleRateInHz * fModAmp;
            m_ppCRingBuffer[i] -> putPostInc(ppfInputBuffer[i][j]);
            ppfOutputBuffer[i][j] = m_ppCRingBuffer[i] -> get(fDelay);
            m_ppCRingBuffer[i] -> getPostInc();
        }
    }
    
    return Error_t::kNoError;
    
}
````
The returned value from CLfo is merely a sinewave data serving as a modulation amplititude. The fraction delay is used in CVibrato::process function.

Reply:
These lines of code are simply getting value from the sinewave, while the fractional index is being used at in CVibrato::process().
```
Vibrato.cpp - Line 125
ppfOutputBuffer[i][j] = m_ppCRingBuffer[i] -> get(fDelay);
```